Raise Error After Failed Cast in GEOS Impls

keithdoggett opened the pull request:
### Summary

Raise errors instead of returning `nil` when objects cannot be properly cast to GEOS geometries in both the capi and FFI implementations.

These methods are primarily called when a method requires a second geometry (ex. `intersects?` takes a `rhs` argument). So now if invalid data is given in `rhs` an error will be raised.

keithdoggett commented on ext/geos_c_impl/factory.c:
```diff
@@ -832,7 +832,8 @@ rgeo_convert_to_geos_geometry(VALUE factory, VALUE obj, VALUE type)
   }
   if (NIL_P(object))
-    return NULL;
+    rb_raise(rb_eRGeoInvalidGeometry,
+             "Unable to cast the geometry to the GEOS Factory");
```
Some tests were failing because they check that an `InvalidGeometry` error is raised rather than any error, so I left the conditional in for now. We can remove it all together and allow `Check_TypedStruct` to raise, but we'll need to fix the tests.

keithdoggett commented on ext/geos_c_impl/factory.c:
```diff
@@ -863,9 +864,11 @@ rgeo_convert_to_detached_geos_geometry(VALUE obj,
+             "Unable to cast the geometry to the GEOS Factory");
   }
 
+  Check_TypedStruct(object, &rgeo_geometry_type);
```
May not be necessary, but added it for safety.

keithdoggett commented on ext/geos_c_impl/geometry.c:
```diff
@@ -421,19 +421,17 @@ method_geometry_disjoint(VALUE self, VALUE rhs)
-      if (prep)
-        val = GEOSPreparedDisjoint(prep, rhs_geom);
-      else
+    prep = rgeo_request_prepared_geometry(self_data);
+    if (prep)
+      val = GEOSPreparedDisjoint(prep, rhs_geom);
+    else
```
I removed all of the checks since we should no longer have a `NULL` `rhs_geom`.

keithdoggett commented on ext/geos_c_impl/geometry_collection.c:
```diff
@@ -66,9 +66,7 @@ create_geometry_collection(VALUE module, int type, VALUE factory, VALUE array)
   for (i = 0; i < len; ++i) {
     geom = rgeo_convert_to_detached_geos_geometry(
       rb_ary_entry(array, i), factory, cast_type, &klass, &state);
-    if (state || !geom) {
-      break;
-    }
```
I think this is ok to remove since `rgeo_convert_to_detached_geos_geometry` will raise on this condition.

keithdoggett commented on ext/geos_c_impl/polygon.c:
```diff
@@ -253,12 +253,6 @@ cmethod_create(VALUE module,
   linear_ring_type = rgeo_feature_linear_ring_module;
   exterior_geom = rgeo_convert_to_detached_geos_geometry(
     exterior, factory, linear_ring_type, NULL, &state);
-  if (state) {
-    rb_exc_raise(rb_errinfo());
-  }
-  if (!exterior_geom) {
-    return Qnil;
-  }
```
Again, I think this is safe to remove because `rgeo_convert_to_detached_geos_geometry` checks for both of these, but I'm not quite sure.

keithdoggett commented on lib/rgeo/geos/ffi_feature_methods.rb:
```diff
@@ -158,11 +158,10 @@ def make_valid
         fg = factory.convert_to_fg_geometry(rhs)
-        if !fg
-          false
+
```
similar situation here. We can remove the `nil` checks

keithdoggett commented on ext/geos_c_impl/geometry_collection.c:
```diff
@@ -64,8 +84,23 @@ create_geometry_collection(VALUE module, int type, VALUE factory, VALUE array)
+    args.klasses = &klass;
+    args.state = &state;
+
+    geom = rb_protect(rgeo_convert_to_detached_geos_geometry_wrapper,
```
Since we're raising now, this function and the `coord_seq_from_array` function had memory leaks because we allocated data before calling `rgeo_convert_*` in each of them. This wraps those calls in `rb_protect` so that we can handle the error, free the existing data and then re-raise.

I could go with `rb_rescue`, but `rb_protect` feels a little nicer to handle all of the error logic within the original caller.

BuonOmo commented on ext/geos_c_impl/factory.c:
```diff
@@ -863,9 +864,11 @@ rgeo_convert_to_detached_geos_geometry(VALUE obj,
                               ID2SYM(rb_intern("keep_subtype")));
   if (*state || NIL_P(object)) {
-    return NULL;
+    rb_raise(rb_eRGeoInvalidGeometry,
+             "Unable to cast the geometry to the GEOS Factory");
```
For (1.), if we are scared of using `rb_set_errinfo` (because `rb_protect` actually seems to do much more), we could:

```c
  if (NIL_P(object))
    rb_protect(rb_exc_raise, rb_exc_new_str("Unable to cast the geometry to the GEOS Factory"), state);

  if (*state)
    return NULL;
```

keithdoggett replied:
I like this solution. We'd have to do a `NULL` check but at least we can raise right then and halt execution instead of letting it propagate.

keithdoggett replied:
I'll look at this soon and see if any other issues come up with it, but this seems like a fairly straightforward fix. Do you suggest doing this for both of the `geos_convert_*` methods or just the `detached` version?

keithdoggett replied:
I guess both of them get used in a similar way that we see as an issue here so I can do both.

keithdoggett replied:
@BuonOmo I'm going through this a little more now and I'm not sure we can meet the guarantee of the `rgeo_convert_*` methods not raising since we use `Check_TypedStruct` to validate the data types which may raise.

I can modify the `detached` version so that if it's `NULL` or `state` is returned, the user should handle the error, and the `state` variable doesn't just disappear, but I don't think it makes sense to pull the `Check_TypedStruct` calls out of these functions. What do you think?

keithdoggett replied:
https://github.com/rgeo/rgeo/commit/6f241060f45d0477e23492794c48e789d0dd3190 This is the commit where we a added that check btw.

keithdoggett replied:
@BuonOmo looked at this some more. It seems like I can replace these `Check_TypedStruct` calls with the `rgeo_is_geos_object` or `rgeo_check_geos_object` functions we defined and wrap it in `rb_protect`.

BuonOmo replied:
@keithdoggett it seems like there was a lot more to dig than what I expected. Maybe raising is the simpler option?

If you feel like you have the problem under control however, I fully agree with your choices.

My last concern is about performances, they may be a bit degraded when using protect. But tbh I don't know, and I'm not sure for now how to handle that.

keithdoggett replied:
Yeah I can look into `rb_protect` and see if performance is going to be really degraded. I think `rb_protect` is intended for calling ruby methods from C code, but the interface is much nicer since you don't have to define a callback function with specific args like you do with `rb_rescue`.

I would say from a "coding flow" standpoint it feels good right now and the 2 `rgeo_convert_*` functions have a clear error handling pattern through `state`, but I can investigate `rb_protect`.

keithdoggett commented:
@BuonOmo my most recent changes propagate `state` through the internal `rgeo_convert_*` functions and the caller must determine how to handle an error. This makes cleanup a lot easier in cases where some data may have already been allocated.

BuonOmo commented on ext/geos_c_impl/factory.c:
```diff
@@ -831,10 +832,21 @@ rgeo_convert_to_geos_geometry(VALUE factory, VALUE obj, VALUE type)
   }
-  if (NIL_P(object))
+  if (NIL_P(object)) {
+    msg = rb_str_new_cstr("Unable to cast the geometry to the GEOS Factory");
+    rb_protect(
+      rb_exc_raise, rb_exc_new_str(rb_eRGeoInvalidGeometry, msg), state);
```
IDK if you had an issue, but you could avoid the temp ruby variable

```suggestion
    rb_protect(
      rb_exc_raise, rb_exc_new_cstr(rb_eRGeoInvalidGeometry, "Unable to cast the geometry to the GEOS Factory"), state);
```

keithdoggett replied:
Ahh completely missed that there's a `cstr` version of the function. I'll change to that. Tried using a cstr in the function I used and it would compile, but segfaulted once it got to that line.

BuonOmo commented on ext/geos_c_impl/factory.c:
```diff
@@ -862,7 +876,18 @@ rgeo_convert_to_detached_geos_geometry(VALUE obj,
-  if (*state || NIL_P(object)) {
+
+  if (NIL_P(object)) {
+    msg = rb_str_new_cstr("Unable to cast the geometry to the GEOS Factory");
```
same here of course!

BuonOmo commented on ext/geos_c_impl/geometry_collection.c:
```diff
@@ -66,9 +66,14 @@ create_geometry_collection(VALUE module, int type, VALUE factory, VALUE array)
+        GEOSGeom_destroy(geoms[j]);
+      }
+      FREE(geoms);
+      rb_jump_tag(state);
```
nice! Maybe we could remove all of my `rb_exc_raise(rb_errinfo())` with that, seems like the way to go!

keithdoggett replied:
I didn't change the ones that were already in there, but I will change to that now.

BuonOmo commented on ext/geos_c_impl/geometry.c:
```diff
@@ -420,20 +421,22 @@ method_geometry_disjoint(VALUE self, VALUE rhs)
-      } else if (val == 1) {
-        result = Qtrue;
-      }
+      val = GEOSDisjoint(self_geom, rhs_geom);
+    if (val == 0) {
+      result = Qfalse;
+    } else if (val == 1) {
+      result = Qtrue;
```
what do you think of

```suggestion

      result = GEOSDisjoint(self_geom, rhs_geom) ? Qtrue : Qfalse
```

BuonOmo reviewed (approved):
Seems like we have something much cleaner now!

keithdoggett commented:
@BuonOmo thanks for the review! The only thing that is a minor concern (although it doesn't seem to be a big deal) is that `rb_protect` expects the given function to return a `VALUE`. The compiler gives a warning about this, but as far as I can tell it doesn't cause any crashes or other issues.

If we want to match the return type, we could wrap `rb_raise`/`rb_exc_raise` in a function like this

```c
VALUE rb_raise_value(VALUE error_type, char* cstr) {
  rb_raise(error_type, cstr);
  return Qnil;
}
```

BuonOmo commented:
I think I'd add the method to avoid having too much warnings. However, not such if that will really silent the error: you might have another error then saying that `rb_raise_value` is not returning and should be declared as such

keithdoggett commented:
@BuonOmo I added a function that returns a `VALUE`. Similar to this use case from the ruby library

https://github.com/ruby/ruby/blob/d92f09a5eea009fa28cd046e9d0eb698e3d94c5c/ext/-test-/exception/ensured.c#L21-L33

I also cleaned up a few other warnings.

BuonOmo commented on ext/geos_c_impl/ruby_more.h:
```diff
@@ -12,6 +12,9 @@ RGEO_BEGIN_C
 VALUE
 rb_protect_funcall(VALUE recv, ID mid, int* state, int n, ...);
 
+VALUE
+rb_exc_raise_value(VALUE exc);
```
Maybe add a short comment here for the reason?
